src/util/types.h

```diff
@@ -8,4 +8,29 @@
 template <class>
 inline constexpr bool ALWAYS_FALSE{false};
 
+/**
+ * Helper to disable implicit copies, but allow explicit ones.
+ *
+ * This can be used on types that are expensive to copy, and cheap to move.
+ * Explicit Copy() or CopyFrom() can be used where slow copies are desired.
+ *
+ * To use, place this as the last line into a class or struct.
+ */
+#define DISABLE_IMPLICIT_COPIES(T)    \
+    T() = default;                    \
+    T& operator=(T&&) = default;      \
+    T(T&&) = default;                 \
+    T Copy() const                    \
+    {                                 \
+        return *this;                 \
+    }                                 \
+    void CopyFrom(const T& from)      \
+    {                                 \
+        *this = from;                 \
+    }                                 \
+                                      \
+private:                              \
+    T& operator=(const T&) = default; \
+    T(const T&) = default
```

Comment:
In commit "Add DISABLE_IMPLICIT_COPIES helper" (fa540a3c5a08c06917a3df1b3d42939eab757f5e)

I think after adding C++20 support might need to drop these T constructors and go back to using the `NoCopyOnlyMove _no_copy_only_move` empty member trick, because according to p1008 if a class has any user-declared constructors it will no longer be an aggregate.

This only applies to constructors, not `operator=` methods, though, so I think private `operator=` method would be able to stay, and `CopyFrom` method would not have to change. `Copy` method might need to change to `T Copy() const { T copy; copy = *this; return copy; }`.

But it might make sense keep this current implementation for now and change it if needed in the C++20 PR later.

Reply:
It looks like an implicitly `deleted` function can not be "recovered" by `default`ing.

```
warning: explicitly defaulted copy assignment operator is implicitly deleted [-Wdefaulted-function-deleted]
error: object of type 'CSerializedNetMsg' cannot be assigned because its copy assignment operator is implicitly deleted
```

Thus, it seems impossible to have a `Copy`/`CopyFrom` macro *and* aggregate initialization that works with both C++17 and C++20.

Reply:
> It looks like an implicitly `deleted` function can not be "recovered" by `default`ing.

I don't think the followup I'm suggesting should cause the copy asssignment operator to be implicitly deleted, so there be no need to recover it. But the PR in its current form seems fine, and we can iterate on it. If iteraton requires dropping explicit Copy methods to support c++20, that seems sad but not tragic.

> Thus, it seems impossible to have a `Copy`/`CopyFrom` macro _and_ aggregate initialization that works with both C++17 and C++20.

I'm not convinced of this, but there's no need to solve it now, and I think the macro here already provides an improvement. It would be good to have a unit test to check that the macro allows all the things it is supposed to allow in c++17. If it has to drop support for some of those things in c++20, the test could catch that and be updated at that time to reflect this.

Reply:
> I'm not convinced of this

Heh, what about you open a pull to prove me wrong?

Reply:
> > I'm not convinced of this
> 
> Heh, what about you open a pull to prove me wrong?

I can try later, but I like this PR because it adds a simple and clean macro for c++17. If the macro has to get messier or stupider later to support c++20, I think it's better if that happens in a followup PR adding c++20 support than in an earlier PR before we are even allowed to use c++20 features.

Reply:
My thinking is that it is possible to achieve without a macro, manually implementing `Copy`/`CopyFrom` as seen here: https://github.com/bitcoin/bitcoin/pull/24169/files#diff-422879cc8bfac56d4380c865f381b58afeb344bc355bbc7f47c581e4491b6b4b

I will rebase this pull after #24169 and probably drop the `Copy`/`CopyFrom` from the macro. If there is a way to achieve them with a macro, it can be done later.